synapse/storage/databases/main/event_federation.py

```diff
@@ -1708,7 +1708,7 @@ def _delete_old_forward_extrem_cache_txn(txn: LoggingTransaction) -> None:
                 WHERE stream_ordering < ?
             """
             txn.execute(
-                sql, (self.stream_ordering_month_ago)  # type: ignore[attr-defined]
+                sql, (self.stream_ordering_month_ago,)  # type: ignore[attr-defined]
```

Comment:
Anyway for the linting to catch this?

Reply:
https://github.com/matrix-org/synapse/blob/6d6fcc98a239dffd564923e96cdec3103af88031/synapse/storage/database.py#L397-L398 could require that `args: Tuple[Any]`  maybe?

Reply:
Oh, that's actually already the case. If `*args: T` then `args: Tuple[T]`. So here `args: Tuple[Any]`.

Tracing it through: `_do_execute_` is annotated to require that `self.txn.execute(sql, *args)` is type-safe.

https://github.com/matrix-org/synapse/blob/bd4d958aaf7c2123abb3665e7a7b199cf8ce27ee/synapse/storage/database.py#L415-L421

`self.txn` is a `Cursor` and `Cursor.execute` is annotated as

https://github.com/matrix-org/synapse/blob/ffc2ee521d26f5b842df7902ade5de7a538e602d/synapse/storage/types.py#L34-L39

So for  `self.txn.execute(sql, *args)` to be type-correct, we need `*args` to contain exactly one argument `arg[0]` which is a `_Parameters` instance. In this situation `arg[0] : Any` so this trivially passes.

I think the signature for `execute` is wrong. I think it should match that of `self.txn.execute` and take a `_Parameters`.

Well, consider me sniped. I'm going to see if I can PR this.

Reply:
-> #15432